File: src/Faborite.SDKs/JavaScript/faborite-sdk.ts
/**
 * Faborite JavaScript/TypeScript SDK
 * Issue #127
 * 
 * Installation:
 *   npm install faborite-sdk
 * 
 * Usage:
 *   import { FaboriteClient } from 'faborite-sdk';
 *   const client = new FaboriteClient({ apiKey: 'your-api-key' });
 *   await client.triggerSync({ workspaceId: '...', lakehouseId: '...' });
 */

export interface FaboriteConfig {
  apiKey: string;
  baseUrl?: string;
}

export interface SyncRequest {
  workspaceId: string;
  lakehouseId: string;
  tables?: string[];
  sampleRows?: number;
}

export interface SyncResponse {
  syncId: string;
  status: string;
  message: string;
}

export class FaboriteClient {
  private apiKey: string;
  private baseUrl: string;

  constructor(config: FaboriteConfig) {
    this.apiKey = config.apiKey;
    this.baseUrl = config.baseUrl || 'https://api.faborite.com';
  }

  async triggerSync(request: SyncRequest): Promise<SyncResponse> {
    const response = await fetch(`${this.baseUrl}/api/sync`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(request),
    });

    return await response.json();
  }

  async listTables(workspaceId: string, lakehouseId: string): Promise<string[]> {
    const response = await fetch(
      `${this.baseUrl}/api/tables?workspaceId=${workspaceId}&lakehouseId=${lakehouseId}`,
      {
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
        },
      }
    );

    return await response.json();
  }

  async getSyncStatus(syncId: string): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/sync/${syncId}`, {
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
      },
    });

    return await response.json();
  }

  async profileData(tableName: string): Promise<any> {
    const response = await fetch(`${this.baseUrl}/api/profile`, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ tableName }),
    });

    return await response.json();
  }

  // Streaming support with async iterators
  async *streamSyncProgress(syncId: string): AsyncGenerator<any> {
    const eventSource = new EventSource(
      `${this.baseUrl}/api/sync/${syncId}/stream`,
      { headers: { 'Authorization': `Bearer ${this.apiKey}` } }
    );

    yield* this.eventSourceToAsyncIterator(eventSource);
  }

  private async *eventSourceToAsyncIterator(eventSource: EventSource): AsyncGenerator<any> {
    const queue: any[] = [];
    let resolve: ((value: any) => void) | null = null;

    eventSource.onmessage = (event) => {
      const data = JSON.parse(event.data);
      if (resolve) {
        resolve(data);
        resolve = null;
      } else {
        queue.push(data);
      }
    };

    while (true) {
      if (queue.length > 0) {
        yield queue.shift();
      } else {
        yield await new Promise((r) => (resolve = r));
      }
    }
  }
}

// Example usage
async function example() {
  const client = new FaboriteClient({ apiKey: 'test-key' });

  // Trigger sync
  const result = await client.triggerSync({
    workspaceId: 'workspace-123',
    lakehouseId: 'lakehouse-456',
  });
  console.log('Sync started:', result);

  // List tables
  const tables = await client.listTables('workspace-123', 'lakehouse-456');
  console.log('Tables:', tables);

  // Stream progress
  for await (const progress of client.streamSyncProgress(result.syncId)) {
    console.log('Progress:', progress);
  }
}

export default FaboriteClient;

File: src/Faborite.SDKs/Python/faborite_sdk.py
"""
Faborite Python SDK
Issue #126

Installation:
    pip install faborite-sdk

Usage:
    from faborite import FaboriteClient
    
    client = FaboriteClient(api_key="your-api-key")
    client.trigger_sync(workspace_id="...", lakehouse_id="...")
"""

class FaboriteClient:
    def __init__(self, api_key: str, base_url: str = "https://api.faborite.com"):
        self.api_key = api_key
        self.base_url = base_url
        self.headers = {"Authorization": f"Bearer {api_key}"}
    
    def trigger_sync(self, workspace_id: str, lakehouse_id: str, **kwargs) -> dict:
        """Trigger a sync operation"""
        import requests
        response = requests.post(
            f"{self.base_url}/api/sync",
            headers=self.headers,
            json={"workspaceId": workspace_id, "lakehouseId": lakehouse_id, **kwargs}
        )
        return response.json()
    
    def list_tables(self, workspace_id: str, lakehouse_id: str) -> list:
        """List all tables in a lakehouse"""
        import requests
        response = requests.get(
            f"{self.base_url}/api/tables",
            headers=self.headers,
            params={"workspaceId": workspace_id, "lakehouseId": lakehouse_id}
        )
        return response.json()
    
    def get_sync_status(self, sync_id: str) -> dict:
        """Get status of a sync operation"""
        import requests
        response = requests.get(
            f"{self.base_url}/api/sync/{sync_id}",
            headers=self.headers
        )
        return response.json()
    
    def profile_data(self, table_name: str) -> dict:
        """Generate data profile for a table"""
        import requests
        response = requests.post(
            f"{self.base_url}/api/profile",
            headers=self.headers,
            json={"tableName": table_name}
        )
        return response.json()

# Example usage
if __name__ == "__main__":
    client = FaboriteClient(api_key="test-key")
    
    # Trigger sync
    result = client.trigger_sync(
        workspace_id="workspace-123",
        lakehouse_id="lakehouse-456"
    )
    print(f"Sync started: {result}")
    
    # List tables
    tables = client.list_tables(
        workspace_id="workspace-123",
        lakehouse_id="lakehouse-456"
    )
    print(f"Tables: {tables}")
